src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -613,5 +667,34 @@ internal ImmutableArray<CompletionProvider> GetAllProviders(ImmutableHashSet<str
                     cancellationToken);
             }
         }
+
+        private class ProjectCompletionProvider
+            : AbstractProjectExtensionProvider<CompletionProvider, ExportCompletionProviderAttribute>
+        {
+            public ProjectCompletionProvider(AnalyzerReference reference)
+                : base(reference)
+            {
+            }
+
+            protected override bool SupportsLanguage(ExportCompletionProviderAttribute exportAttribute, string language)
+            {
+                return exportAttribute.Language == null
+                    || exportAttribute.Language.Length == 0
+                    || exportAttribute.Language.Contains(language);
+            }
+
+            protected override bool TryGetExtensionsFromReference(AnalyzerReference reference, out ImmutableArray<CompletionProvider> extensions)
+            {
+                // check whether the analyzer reference knows how to return completion providers directly.
+                if (reference is ICompletionProviderFactory completionProviderFactory)
```

Comment:
How does this (and other provider factory) interface work? Who's gonna implement it?

Reply:
One of the test providers implements it

Reply:
Are those interfaces intended for non-test scenario as well?

Reply:
No. I just followed an existing pattern from the original work to handle diagnostic analyzers. Perhaps the pattern could be removed in the future, but it didn't seem necessary to do that here.